configcatclient/rolloutevaluator.py

```diff
@@ -48,174 +49,616 @@ def evaluate(self, key, user, default_value, default_variation_id, settings):  #
+            except ValueError:
+                validation_error = "'%s' is not a valid semantic version" % str(user_value).strip()
+                error = self._handle_invalid_user_attribute(comparison_attribute, comparator, comparison_value, key,
+                                                            validation_error)
+                return False, error
+        # =, <>, <, <=, >, >= (number)
+        elif Comparator.EQUALS_NUMBER <= comparator <= Comparator.GREATER_THAN_OR_EQUAL_NUMBER:
+            try:
+                user_value_float = float(str(user_value).replace(",", "."))
+            except ValueError:
+                validation_error = "'%s' is not a valid decimal number" % str(user_value)
+                error = self._handle_invalid_user_attribute(comparison_attribute, comparator, comparison_value, key,
+                                                            validation_error)
+                return False, error
+
+            comparison_value_float = float(str(comparison_value).replace(",", "."))
+
+            if (comparator == Comparator.EQUALS_NUMBER and user_value_float == comparison_value_float) \
+                    or (comparator == Comparator.NOT_EQUALS_NUMBER and user_value_float != comparison_value_float) \
+                    or (comparator == Comparator.LESS_THAN_NUMBER and user_value_float < comparison_value_float) \
+                    or (comparator == Comparator.LESS_THAN_OR_EQUAL_NUMBER and user_value_float <= comparison_value_float) \
+                    or (comparator == Comparator.GREATER_THAN_NUMBER and user_value_float > comparison_value_float) \
+                    or (comparator == Comparator.GREATER_THAN_OR_EQUAL_NUMBER and user_value_float >= comparison_value_float):
+                return True, error
+        # IS ONE OF (hashed)
+        elif comparator == Comparator.IS_ONE_OF_HASHED:
+            if sha256(encode_utf8(user_value), salt, context_salt) in comparison_value:
+                return True, error
+        # IS NOT ONE OF (hashed)
+        elif comparator == Comparator.IS_NOT_ONE_OF_HASHED:
+            if sha256(encode_utf8(user_value), salt, context_salt) not in comparison_value:
+                return True, error
+        # BEFORE, AFTER (UTC datetime)
+        elif Comparator.BEFORE_DATETIME <= comparator <= Comparator.AFTER_DATETIME:
+            try:
+                user_value_float = float(str(user_value).replace(",", "."))
+            except ValueError:
+                validation_error = "'%s' is not a valid Unix timestamp (number of seconds elapsed since Unix epoch)" % \
+                                   str(user_value)
+                error = self._handle_invalid_user_attribute(comparison_attribute, comparator, comparison_value, key,
+                                                            validation_error)
+                return False, error
+
+            comparison_value_float = float(str(comparison_value).replace(",", "."))
```

Comment:
Conversion to string looks unnecessary. We should receive a number from a valid config json.